release-notes/release-notes.go

```diff
@@ -18,130 +18,82 @@ package main
 
 import (
 	"bytes"
+	goflag "flag"
 	"fmt"
-	"net/http"
 	"os"
-	"sort"
 	"time"
 
-	"github.com/google/go-github/github"
-	flag "github.com/spf13/pflag"
-	"golang.org/x/oauth2"
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	"k8s.io/contrib/mungegithub/github"
 )
 
 var (
-	base    string
-	last    int
-	current int
-	token   string
+	base        string
+	last        int
+	lastTime    time.Time
+	current     int
+	currentTime time.Time
 )
 
-type byMerged []*github.PullRequest
-
-func (a byMerged) Len() int           { return len(a) }
-func (a byMerged) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
-func (a byMerged) Less(i, j int) bool { return a[i].MergedAt.Before(*a[j].MergedAt) }
-
-func init() {
-	flag.IntVar(&last, "last-release-pr", 0, "The PR number of the last versioned release.")
-	flag.IntVar(&current, "current-release-pr", 0, "The PR number of the current versioned release.")
-	flag.StringVar(&token, "api-token", "", "Github api token for rate limiting. Background: https://developer.github.com/v3/#rate-limiting and create a token: https://github.com/settings/tokens")
-	flag.StringVar(&base, "base", "master", "The base branch name for PRs to look for.")
+func addFlags(cmd *cobra.Command) {
+	cmd.Flags().IntVar(&last, "last-release-pr", 0, "The PR number of the last versioned release.")
+	cmd.Flags().IntVar(&current, "current-release-pr", 0, "The PR number of the current versioned release.")
+	cmd.Flags().StringVar(&base, "base", "master", "The base branch name for PRs to look for.")
+	cmd.PersistentFlags().AddGoFlagSet(goflag.CommandLine)
 }
 
-func usage() {
-	fmt.Printf(`usage: release-notes --last-release-pr=<number> --current-release-pr=<number>
-                     --api-token=<token> [--base=<branch-name>]
-`)
+func mergedAt(num int, config *github.Config) time.Time {
+	obj, err := config.GetObject(num)
+	if err != nil {
+		glog.Fatalf("Unable to get munge object for %d: %v", num, err)
+	}
+	t := obj.MergedAt()
+	if t == nil {
+		glog.Fatalf("Unable to get merge time for %d: %v", num, err)
+	}
+	return *t
 }
 
 func main() {
-	flag.Parse()
-	if last == 0 || current == 0 || token == "" {
-		usage()
-		os.Exit(1)
-	}
+	config := &github.Config{}
+	cmd := &cobra.Command{
+		Use: "release-notes --last-release-pr=NUM --current-release-pr=NUM",
+		RunE: func(_ *cobra.Command, _ []string) error {
+			if err := config.PreExecute(); err != nil {
+				return err
+			}
+			if last == 0 || current == 0 {
+				return fmt.Errorf("Must set (at least) --last-releae-pr and --current-release-pr")
+			}
+			lastTime = mergedAt(last, config)
+			currentTime = mergedAt(current, config)
 
-	var tc *http.Client
-	if len(token) > 0 {
-		tc = oauth2.NewClient(
-			oauth2.NoContext,
-			oauth2.StaticTokenSource(
-				&oauth2.Token{AccessToken: token}),
-		)
-	}
-	client := github.NewClient(tc)
+			buffer := &bytes.Buffer{}
+			config.ForEachIssueDo("closed", []string{"release-note"}, func(obj *github.MungeObject) error {
```

Comment:
Hm.  I'd like to keep the possibility of listing all PRs, not just those with the release-note label.  With this implementation, isn't that going to require pulling all ~20,000 issues (minus the open ones)?  That seems excessive to me.

In the current implementation, we pull PRs based on when they were last updated, and stop pulling when we see an issue that was last updated before the `last-release-pr` was merged, (and thus couldn't have been merged after).  Could we do something like that here?

(It requires making the `ForEachIssueDo` abstraction a bit more complicated ... I'm not sure if it's worth it.)